File: examples/typescript/order-processing-state/order/order.test.ts
import { Order } from "./order";

function assert(condition: boolean, message: string): void {
  if (!condition) {
    throw new Error(message);
  }
}

function assertThrows(run: () => void, expectedMessage: string): void {
  try {
    run();
    throw new Error("expected function to throw");
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    assert(message.includes(expectedMessage), `expected error to include "${expectedMessage}", got "${message}"`);
  }
}

const order = Order.create("ord-ts-3001");
order.allocateInventory();
order.pack();
order.ship();

assert(order.state() === "shipped", `expected shipped state, got ${order.state()}`);
assert(order.events().length === 3, `expected 3 events, got ${order.events().length}`);

const pending = Order.create("ord-ts-3002");
assertThrows(() => pending.ship(), "cannot ship while order is pending");
assert(pending.state() === "pending", `expected pending state, got ${pending.state()}`);

console.log("order-processing-state tests passed");

File: examples/go/order-processing-state/order/order_test.go
package order

import "testing"

func TestHappyPathTransitionsAndEvents(t *testing.T) {
	order := New("ord-3001")

	if err := order.AllocateInventory(); err != nil {
		t.Fatalf("allocate inventory: %v", err)
	}

	if err := order.Pack(); err != nil {
		t.Fatalf("pack: %v", err)
	}

	if err := order.Ship(); err != nil {
		t.Fatalf("ship: %v", err)
	}

	if got := order.State(); got != "shipped" {
		t.Fatalf("expected shipped state, got %s", got)
	}

	events := order.Events()
	if len(events) != 3 {
		t.Fatalf("expected 3 events, got %d", len(events))
	}
}

func TestInvalidTransitionFromPending(t *testing.T) {
	order := New("ord-3002")

	if err := order.Ship(); err == nil {
		t.Fatal("expected shipping from pending to fail")
	}

	if got := order.State(); got != "pending" {
		t.Fatalf("expected pending state, got %s", got)
	}
}
